### ISSUE TITLE
How get information about polygon?

### ISSUE THREAD

**Issue by 0xtrzy** [boostorg/geometry#728]
How get information about:
- number of points?
- number of segments?
- clockwise or not?

this is my code https://pastebin.com/R9niv31u

-----
typedef boost::geometry::model::d2::point_xy<double> point_t;
typedef boost::geometry::model::polygon<point_t> polygon_t;
....
this working:
std::cout << "czy jest prosta?   " << (boost::geometry::is_simple(green) ? "tak" : "nie") 
this dont.
(boost::geometry::clockwise(green)

Meybe I must using Multi polygon?

**Comment by mloskot** [comment]
The `boost::geometry::clockwise` is not a function, it's enumerator.

https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/enumerations/order_selector.html
https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/core/point_order.html

1. Please, try to read the docs!
2. Please, try to write in English.
3. Please, try to learn about how to format code blocks in Markdown.

**Comment by 0xtrzy** [comment]
my english are ugly ;( 

but this example not showing how determining which polygon are clockwise.

**Comment by mloskot** [comment]
It's a static property of a type:
https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/concepts/concept_ring.html

If you run `correct`, it will fix any unordered points, according to that type trait
https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/algorithms/correct/correct_1.html

You may also try `is_valid` to see if the message will give you extra details on unordered points, I can't remember if it does
https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/algorithms/is_valid/is_valid_2_with_message.html

**Comment by 0xtrzy** [comment]
correct change whole polygon (with more than one shape)
sometime are not a good idea. For example I have 'A' shape. I must have outer and hole inside. I need set exactly one shape to clockwise and other to not clockwise.

is_valid exist in my code but show me a error? Why showing is not a valid? This is simple shape only 4 point.

**Comment by mloskot** [comment]
If you read my https://github.com/boostorg/geometry/issues/728#issuecomment-645991346 carefully and used `is_valid` as I suggested, you would have answered your own question:


```cpp
#include <iostream>
#include <string>

#include <boost/geometry.hpp>
#include <boost/geometry/geometries/point_xy.hpp>
#include <boost/geometry/geometries/polygon.hpp>
#include <boost/geometry/io/wkt/wkt.hpp>

typedef boost::geometry::model::d2::point_xy<double> point_t;
typedef boost::geometry::model::polygon<point_t> polygon_t;

int main()
{
    polygon_t green;
    boost::geometry::read_wkt("POLYGON((0 4, 3 1, 6 4, 3 0)(0 3, 0.1 3, 0.1 3.5, 0 3.5))", green);  

    std::string msg;
    std::cout << "is valid? " << boost::geometry::is_valid(green, msg) << std::endl;  
    std::cout << msg << std::endl;
}
```
Outputs
```
is valid? 0
Geometry is defined as closed but is open
```
Try it yourself https://godbolt.org/z/d-Zb9Y

**Comment by 0xtrzy** [comment]
Yes, my code use this same way. in 59 line

Question is how check one shape, (sub-polygon?)
and how setup: clockwise or dont clockwise.

and number of points is in code but comment because not working.

**Comment by mloskot** [comment]
In the docs that I linked in https://github.com/boostorg/geometry/issues/728#issuecomment-645991346 about polygon and ring, you can read:

> If the polygons underlying ring_type is defined as closed, all rings must be closed: the first point must be spatially equal to the last point. 

Rings of your polygons are not closed. A ring that is not closed is not a valid ring. Period.
So, no, your "shape"-s are not correct according to concepts of Boost.Geometry. Period.

**Comment by 0xtrzy** [comment]
@mloskot in first question I ask: Meybe I must using multi polygon?

I have object similar chars 'A' 'i' 'Ń' etc. My polygon have more than one shape and holes. How join (union) two shapes? How setup hole? How set clockwise. And how get element in this shapes. 

?

**Comment by mloskot** [comment]
Create polygon by creating rings one by one, one exterior and one or more interior rings:

https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/algorithms/append.html

Run `correct` which should fix ordering of points in the rings.

Once again, browse the docs, search for familiar names, discover the library. If you feel lazy, ask on StackOverflow.

**Comment by 0xtrzy** [comment]
I'm not lazy, I 'm not undestand. 

I ask 3 question, You answer only on correct. No example how get point
for example a[1] not working ?
how turn polygon clockwise etc.

and ... closing discusion. Please help me

**Comment by mloskot** [comment]
I did not say you are lazy. I said "if you feel lazy", as every one does from time to time...

I did answer all your questions

-  number of points?

> Please, read the docs! Hint: `num_*`

In other words, search for the function https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/algorithms/num_points.html

-  number of segments?

Likewise, https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/algorithms/num_segments.html

-  clockwise or not?

`point_order` to see what order types define, then `correct` to fix it, then you can assume you `point_order` reports valid order for your type.

----

You are asking basic questions which well covered by the library docs.
Authors of the library put a lot of effort in the docs. 
Although there is lots of places where the docs could be improved, I don't think there is anything missing to answer your questions just by browsing and reading the docs.
Basically, answer to your question boils down to RTFM!

**Comment by awulkiew** [comment]
Thanks Mateusz for handling this :)

@0xtrzy The library gets the information about the order (Clockwise vs Counterclockwise) and Closedness (Last point equal to the first point or not) from the polygon type. The data stored inside a polygon as to be consistent with this type definition. See e.g.:
https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/models/model_polygon.html
There are template parameters `ClockWise` and `Closed` defining this. And points stored in the polygon has to be consistent with these template parameters. Like that:
```
namespace bg = boost::geometry;
typedef bg::model::point<double, 2, bg::cs::cartesian> point;
bg::model::polygon<point, true, true> polygon_cw_closed;
bg::model::polygon<point, true, false> polygon_cw_opened;
bg::model::polygon<point, false, true> polygon_ccw_closed;
bg::model::polygon<point, false, false> polygon_ccw_opened;

boost::geometry::read_wkt("POLYGON((0 0, 0 3, 3 3, 3 0, 0 0)(1 1, 2 1, 2 2, 1 2, 1 1))", polygon_cw_closed);
boost::geometry::read_wkt("POLYGON((0 0, 0 3, 3 3, 3 0)(1 1, 2 1, 2 2, 1 2))", polygon_cw_opened);
boost::geometry::read_wkt("POLYGON((0 0, 3 0, 3 3, 0 3, 0 0)(1 1, 1 2, 2 2, 2 1, 1 1))", polygon_ccw_closed);
boost::geometry::read_wkt("POLYGON((0 0, 3 0, 3 3, 0 3)(1 1, 1 2, 2 2, 2 1))", polygon_ccw_opened);
```
Note that interior rings has to have the opposite orientation than the exterior ring. Read about WKT representation: https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry

If you're not sure what was the order of the points in WKT string you can make the data consistent with the template parameters by calling `bg::correct()`, e.g.:
```
// ccw, opened WKT but cw closed type
boost::geometry::read_wkt("POLYGON((0 0, 3 0, 3 3, 0 3)(1 1, 1 2, 2 2, 2 1))", polygon_cw_closed);
bg::correct(polygon_cw_closed);
std::cout << bg::wkt(polygon_cw_closed) << std::endl;
```
In all cases `bg::is_valid(polygon_xxx)` should return true before you can pass it into any algorithm.

You can also learn about the standard Boost.Geometry is conformant with:
https://www.ogc.org/standards/sfa

**Comment by awulkiew** [comment]
Also note that `num_points` will return different value depending on the `Closed`/Opened` definition. 

Furthermore there is additional parameter in `num_points` defining whether or not the function should add 1 to the returned value for opened points in order to always get the same result whether or not the polygon is opened or not. If you pass `true` as the `add_for_open` you'll get the number of points as if the polygon was closed (so had rings with last point equal to the first one). See:
https://www.boost.org/doc/libs/1_73_0/libs/geometry/doc/html/geometry/reference/algorithms/num_points.html

And as Mateusz said, your polygon has to be valid. So call `bg::correct()` if you're not sure.

**Comment by mloskot** [comment]
@0xtrzy Please, close the issue if your questions have been answered.
